Docs 1826 restart impaired vs unresponsive workers hn (#234)
mulesoft/docs-runtime-manager: modules/ROOT/pages/worker-monitoring.adoc

= Application Monitoring and Automatic Restarts
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:keywords: cloudhub, monitoring, worker, restarts

image:logo-cloud-active.png[link="/runtime-manager/deployment-strategies", title="CloudHub"]
image:logo-hybrid-disabled.png[link="/runtime-manager/deployment-strategies", title="Hybrid"]
image:logo-server-disabled.png[link="/runtime-manager/deployment-strategies", title="Anypoint Platform PCE"]
image:logo-rtf-disabled.png[link="/runtime-fabric/deploy-to-runtime-fabric", title="Runtime Fabric"]

CloudHub monitors all applications and enables you to receive notifications and diagnostic information if your application becomes unresponsive.

You can enable automatic restarts so that your application can recover automatically without your intervention.

== How Application Monitoring Works

Monitoring works differently depending on your Mule version.

* Mule 4.x
+
Every 30 seconds, CloudHub checks the PID of the Mule process.
+

> If CloudHub detects the application process is not running three times in a row, the logs print a warning.

Reviewer: ```suggestion
If CloudHub detects that the application process is not running three times in a row, the logs print a warning.
```
Should we clarify "three times in a row"? What time span are we looking at here? or should we say "After an application stops running three times, CloudHub sends a warning to the application logs."

* Mule 3.8.x
+
When you deploy, CloudHub inserts a monitoring ping flow into your application.
+
When the application starts, the logs display a message to indicate that monitoring is enabled and that the ping flow has been injected into your application:
+
.The arrow shows the `Monitoring enabled` and `Registering ping flow injector` messages in the logs.
image::cloudhub-app-monitor-enabled.png[Monitoring enabled and Registering ping flow injector message in the logs]
+

> Every 30 seconds, CloudHub sends an HTTP GET request to the ping flow within the application.

Reviewer: ```suggestion
Every 30 seconds, CloudHub sends an HTTP GET request to the ping flow.
```
Just to avoid the repetition.

+
If the application is running normally, the ping flow sends a response.

> If CloudHub receives no response from the ping three times in a row, the logs display a warning and a full thread dump.

Reviewer: ```suggestion
If CloudHub receives no response from three consecutive pings, the logs display a warning and a full thread dump.
```

Reviewer: Do the logs display the dump or do we mean to say that the dump occurs...or both?

In both cases, CloudHub triggers a notification to the console to alert you that a worker for an application stopped responding.

> If you configure email alerts, the non-responsive worker also triggers an email alert based on the `Worker Not Responding` event.

Reviewer: ```suggestion
If you configure email alerts, the unresponsive worker also triggers an email alert based on the `Worker Not Responding` event.
```

== Enable Automatic Restarts

You can enable automatic restarts when you deploy a new application or redeploy an existing application.

> * Deploy a new app:

Reviewer: Don't need colons in list hierarchies
```suggestion
* Deploy a new app
```

.. In Runtime Manager, click *Applications* in the menu on the left.
.. Click *Deploy Application* and select *CloudHub* from *Deployment Target*.
.. Select the *Automatically restart application when not responding* option:
+
.The arrow shows the *Automatically restart application when not responding* option.
image::cloudhub-app-autorestart.png[Automatically restart application when not responding option]
.. Click *Deploy Application*.

> * Redeploy an existing app:

Reviewer: ```suggestion
* Redeploy an existing app
```

.. In Runtime Manager, click *Applications* in the menu on the left.
.. Click the app status to display the details pane on the right
.. Click the *Manage Application* button to display the *Settings* page for the app.
.. Select the *Automatically restart application when not responding* option:
+
.The arrow shows the *Automatically restart application when not responding* option.
image::cloudhub-app-autorestart.png[Automatically restart application when not responding option]
.. Click *Apply Changes*.

== How Automatic Restarts Work

> * If you select the *Automatically restart application when not responding* option, CloudHub monitors the worker and also automatically restarts the application if necessary.

Reviewer: Consider plain text for the two scenarios, rather than two list items.

+
CloudHub displays a notification that the app is restarting and another to report the success or failure of the restart.
+

> ** If the restart succeeds, the ping script is reinitialized so that monitoring continues.

Reviewer: ...then these become Level-1 bullets.

> ** If the restart fails, CloudHub continues to attempt restarting the app up to five times.

Reviewer: 5 total times or 4 more times?

+

> Once the maximum number of restart attempts is exceeded, CloudHub takes no further action. The monitoring script exits and the application is left unchanged until you manually intervene.

Reviewer: Once > After

+
The logs report the details of the restart procedure.

> * If you don't select  the *Automatically restart application when not responding* option, the monitoring script exits and the application is left unchanged until you manually restart.

Reviewer: extra space
```suggestion
* If you don't select the *Automatically restart application when not responding* option, the monitoring script exits and the application is left unchanged until you manually restart.
```
